### PR TITLE
feat(highlight): support highlight namespaces per window

### PR DESCRIPTION
**Pull request by bfredl**
The test will be if we can reimplement 'winhl' entirely lua side based on temas.

### CODE REVIEW

**Comment by bfredl** [comment]
Demonstration (pick your two favorite color schemes):
```
local a = vim.api
_G.a = vim.api

vim.cmd [[ color wip ]]
atema = a.nvim_create_namespace("hösttema")
adata = a.nvim__get_hl_defs(0)
for k,v in pairs(adata) do
  a.nvim_set_hl(atema, k, v)
end

vim.cmd [[ color gruvbox ]]
btema = a.nvim_create_namespace("jultema")
bdata = a.nvim__get_hl_defs(0)
for k,v in pairs(bdata) do
  a.nvim_set_hl(btema, k, v)
end

vim.cmd [[ color wip ]]

a.nvim_win_set_config(0, {hl_ns=btema})
split
```

File: src/nvim/api/vim.c [hunk 1]
```diff
@@ -250,32 +249,37 @@ void nvim_set_hl(Integer ns_id, String name, Dictionary val, Error *err)
...
+/// Set active namespace for highlights.
+///
+/// NB: this function can be called from async contexts, but the
+/// semantics are not yet well-defined. To start with
+/// |nvim_set_decoration_provider| on_win and on_line callbacks
+/// are explicitly allowed to change the namespace during a redraw cycle.
+///
+/// @param ns_id the namespace to activate
+/// @param[out] err Error details, if any
+void nvim_set_hl_ns_fast(Integer ns_id, Error *err)
```
Comment (justinmk) [inline src/nvim/api/vim.c:275]: don't think we should start adding "fast" suffix to api names. It's already covered by an attribute and could be part of the metadata if it isn't already.

File: src/nvim/api/vim.c [hunk 2]
```diff
@@ -250,32 +249,37 @@ void nvim_set_hl(Integer ns_id, String name, Dictionary val, Error *err)
 /// |nvim_set_decoration_provider| on_win and on_line callbacks
 /// are explicitly allowed to change the namespace during a redraw cycle.
 ///
-/// @param ns_id the namespace to activate
+// @param ns_id the namespace to activate
 /// @param[out] err Error details, if any
-void nvim__set_hl_ns(Integer ns_id, Error *err)
-  FUNC_API_FAST
+void nvim_set_hl_ns(Integer ns_id, Error *err)
```
Comment (justinmk) [inline src/nvim/api/vim.c:254]: should we add a `opt` param now that this is promoted to public api?

**Comment by luukvbaal** [comment]
https://github.com/neovim/neovim/issues/19823

**Pull request merged**